packages/vite/src/node/config.ts

```diff
@@ -589,6 +587,166 @@ export type ResolvedConfig = Readonly<
   } & PluginHookUtils
 >
 
+// inferred ones are omitted
+export const configDefaults = Object.freeze({
```

Comment:
There's many inferred options. Is it fine to simply omit it?

Reply:
I was initially thinking that we'd only export the defaults that're special, like hardcoded array of values. That way we solve the problem of users wanting to extend it rather than completely replacing it.

I'm not sure yet about exposing all the defaults in the object. It seems to be a lot to keep track of, and like you mention here, there's some that's inferred.

Reply:
I'm fine with omitting the inferred ones, maybe adding a comment about what how that inference works. My vote goes for the current implementation in the PR. I think once the "merge" helper is in, it is going to be easier to work with.

Reply:
What kind of merge helper are you referring to? What I guess I'm concerned about with exposing tons of options this way is that people could unknowingly used it like `vite.build(configDefaults)` or use it to merge entire configs which could be wasteful or lead to incorrect merges (like arrays). Hence I was thinking to be a bit conservative here 🤔

Reply:
I was thinking on the `mergeWithDefaults` that Sapphi implemented. We could prevent `vite.build(configDefaults)` by checking the instance. But you're right that they may do something like: `vite.build({ ...configDefaults, foo: 'bar' })`.

I don't think that current users would ever try to do something like this, but depending how we document `configDefaults`, I see your point.

It feels a bit cumbersome to use if we expose a ton of variables like `CONFIG_DEFAULTS_RESOLVE_CONDITIONS`. Imagine having to import 3 or 4 of these.

Maybe we could make all the properties in the public `configDefaults` non-enumerable so they can not iterate over the keys and the user needs to access them one by one. We should get the same properties as the individual variables but with better typing and easier to import. We would still have a `internalConfigDefaults` that we can iterate to be used in `mergeWithDefaults`

Reply:
I'm fine with trimming down the values for the exposed object. I think the object form makes it easier for users to find out the value for each options. But given that we already need to expose `mainFields` and `conditions` separately, maybe that advantage is gone.

Reply:
`mergeWithDefaults` makes sense to me, I think we needed that elsewhere regardless of this PR too.

I guess given:
1. We couldn't expose all defaults as some are inferred.
2. Some defaults are based on the client / server consumer (`mainFields` and `conditions`), which Sapphi has now refactored as `DEFAULT_CLIENT_MAIN_FIELDS` etc (https://github.com/vitejs/vite/pull/18550#discussion_r1830293791)
3. And personally I think I'm missing the usecase where users would want to know the defaults via `configDefaults`. Isn't the problem we want to solve here that if someone wants to add a new condition on `resolve.conditions` but want to extend from Vite defaults, they can use this new export? What are the usecase for e.g. `configDefaults.base`?

I think I'm leaning more on simply exporting the constants like what Sapphi has now with `DEFAULT_CLIENT_MAIN_FIELDS`. We only export the ones that solves the usecase I mentioned in no3.

Reply:
Maybe it is a good idea to merge only the refactoring that makes things cleaner internally for now, and then we can discuss how to expose the defaults in another PR. Maybe we don't need to do this for Vite 6, people can still copy and past the defaults from the docs (it isn't the same though, that will fix the values, instead of adding an extra to the defaults that is what we have now but it may be ok)

Reply:
I pushed a commit that removes the new exports so that we can separate the discussion about exposing the default values. 👍